pkg/controller/windowsmachine/metrics/metrics.go

```diff
@@ -0,0 +1,196 @@
+		}
+		return nil
+	}
+	log.Info("address added to endpoints object", "IPAddress", pc.endpointAddress.IP)
+	return nil
+}
+
+// removeEndpointAddress removes node ip address to the list of subsets in Endpoints object
+func (pc *prometheusNodeConfig) removeEndpointAddress() error {
+	nodeIPPosition, err := pc.isNodeInEndpoints()
+	// If node not found in the endpoints list, return without proceeding further
+	if err != nil {
+		return nil
+	}
+	patchData := fmt.Sprintf("[{\"op\": \"remove\", \"path\": \"/subsets/0/addresses/%s\"}]", nodeIPPosition)
+
+	_, err = pc.k8sclientset.CoreV1().Endpoints(windowsMetricsEndpointsNamespace).Patch(context.TODO(),
+		windowsMetricsEndpoints, types.JSONPatchType, []byte(patchData), metav1.PatchOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "could not remove node %v from endpoints object", pc.hostname)
+	}
+	log.Info("address removed from endpoints object", "IPAddress", pc.endpointAddress.IP)
+	return nil
+}
+
+// addMetrics will create the Services and Service Monitors to allow the operator export the metrics by using
+// the Prometheus operator
+func AddMetrics(ctx context.Context, cfg *rest.Config, namespace string) {
+
+	// Add to the below struct any other metrics ports you want to expose.
+	servicePorts := []v1.ServicePort{
+		{Port: Port, Name: metricsPortName, Protocol: v1.ProtocolTCP, TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: Port}},
+	}
+
+	// Create Service object to expose the metrics port(s).
+	service, err := metrics.CreateMetricsService(ctx, cfg, servicePorts)
+	if err != nil {
+		log.Error(err, "could not create metrics Service")
+	}
+	oclient, err := k8sclient.NewForConfig(cfg)
+	if err != nil {
+		log.Error(err, "could not create config clientset")
+	}
+
+	// When a selector is present in a headless service i.e. spec.ClusterIP=None, Kubernetes manages the
+	// list of endpoints reverting all the changes made by the operator. Remove selector from Metrics Service to avoid
+	// reverting changes in the Endpoints object.
+	patchData := fmt.Sprintf(`{"spec":{"selector": null }}`)
+	service, err = oclient.CoreV1().Services(namespace).Patch(ctx, service.Name, types.MergePatchType,
+		[]byte(patchData), metav1.PatchOptions{})
+	if err != nil {
+		log.Error(err, "could not remove selector from metrics service")
+	}
+
+	// CreateServiceMonitors will automatically create the prometheus-operator ServiceMonitor resources
+	// necessary to configure Prometheus to scrape metrics from this operator.
+	services := []*v1.Service{service}
+	_, err = metrics.CreateServiceMonitors(cfg, namespace, services)
+	if err != nil {
+		log.Info("could not create ServiceMonitor object", "error", err.Error())
+		// If this operator is deployed to a cluster without the prometheus-operator running, it will return
+		// ErrServiceMonitorNotPresent, which can be used to safely skip ServiceMonitor creation.
```

Comment:
In this case won't the endpoint update also fail causing node configuration to always be retried but never succeed? I think you need to figure out if metrics is enabled before proceeding with `Configure()`.

Reply:
I can check if `ServiceMonitor` exists in Configure()